Mathlib/Tactic/Linarith/SimplexAlgo/Gauss.lean

```diff
@@ -0,0 +1,125 @@
+/-
+Copyright (c) 2024 Vasily Nesterov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Vasily Nesterov
+-/
+import Mathlib.Tactic.Linarith.SimplexAlgo.Datatypes
+
+/-!
+# Gaussian Elimination algorithm
+
+The first step of `Linarith.SimplexAlgo.findPositiveVector` is finding initial feasible solution
+which is done by standard Gaussian Elimination algorithm implemented in this file.
+-/
+
+namespace Linarith.SimplexAlgo.Gauss
+
+/-- The mutable state for the `GaussM` monad. -/
+structure GaussState (n m : Nat) where
+  /-- Current matrix. -/
+  mat : Matrix n m
+  /-- Current row. -/
+  currentRow : Nat
+  /-- Current column. -/
+  currentColumn : Nat
+
+/-- The monad for the Gaussian Elimination algorithm. -/
+abbrev GaussM (n m : Nat) := StateM <| GaussState n m
+
+/-- Gets current row. -/
+def curRow {n m : Nat} : GaussM n m Nat := do
+  return (← get).currentRow
+
+/-- Gets current column. -/
+def curCol {n m : Nat} : GaussM n m Nat := do
+  return (← get).currentColumn
+
+/-- Increments current row. -/
+def incRow {n m : Nat} : GaussM n m Unit :=
+  modify fun s => {s with currentRow := s.currentRow + 1}
+
+/-- Increments current column. -/
+def incCol {n m : Nat} : GaussM n m Unit :=
+  modify fun s => {s with currentColumn := s.currentColumn + 1}
+
+/-- Finds the first row starting from the current column with nonzero element in current column. -/
+def findNonzeroRow {n m : Nat} : GaussM n m <| Option Nat := do
+  for i in [← curRow:n] do
+    if (← get).mat[i]![← curCol]! != 0 then
+      return i
+  return .none
+
```

Comment:
I'm not convinced this `GaussM` monad is pulling its weight here; I'd be inclined to drop all these row operations, and:

* Make `findNonzeroRow` take `curRow` and `curCol`
* Use `let mut row` and `let mut col` in `getTableImp`

Happy to let another reviewer override me.